### PR TITLE
Correctly inject config in `PytorchModelHubMixin`

### PR DESCRIPTION
**Pull request by Wauplin**
Solves https://github.com/huggingface/huggingface_hub/pull/2079.

With this PR, users don't have to rely on a  `config` parameter to configure their model. We read default+passed values automatically to build the config file ourselves. This way, we are able to save a correct `config.json` file so that reloaded model have exact same config. Only "jsonable" fields are saved in config.json.

And ofc it's backward compatible so users that prefer to pass `config.json` directly are still supported.

PR is inspired by Inspired by https://github.com/facebookresearch/hiera/pull/26. cc @dbolya with this change, you won't need the `@has_config` decorator anymore :hugs: 

Here is an example to showcase it:

```python
>>> import torch
>>> import torch.nn as nn
>>> from huggingface_hub import PyTorchModelHubMixin

>>> class MyModel(nn.Module, PyTorchModelHubMixin):
...     def __init__(self, hidden_size: int = 512, vocab_size: int = 30000, output_size: int = 4):
...         super().__init__()
...         self.param = nn.Parameter(torch.rand(hidden_size, vocab_size))
...         self.linear = nn.Linear(output_size, vocab_size)

...     def forward(self, x):
...         return self.linear(x + self.param)
>>> model = MyModel(hidden_size=256)

# Save model weights to local directory
>>> model.save_pretrained("my-awesome-model")

# Push model weights to the Hub
>>> model.push_to_hub("my-awesome-model")

# Download and initialize weights from the Hub
>>> model = MyModel.from_pretrained("username/my-awesome-model")
>>> model.hidden_size
256
```

cc @mfuntowicz  as well. This shouldn't impact you except that the generated config.json files should be more complete.

### CODE REVIEW

**Comment by NielsRogge** [comment]
That looks great. However, in your PR description, do you mean `self.hidden_size` instead of `config.hidden_size`?
```diff
self.param = nn.Parameter(torch.rand(self.hidden_size, config.vocab_size))
self.linear = nn.Linear(self.output_size, config.vocab_size)

**Review by NielsRogge — APPROVED** [review]
Great! This is going to be very useful

**Comment by Wauplin** [comment]
> That looks great. However, in your PR description, do you mean self.hidden_size instead of config.hidden_size etc. in order to access the attributes?

Yes sorry, I've updated the example.

File: src/huggingface_hub/hub_mixin.py [hunk 1]
```diff
@@ -428,26 +481,19 @@ class PyTorchModelHubMixin(ModelHubMixin):
...
-    ...     hidden_size: int = 512
-    ...     vocab_size: int = 30000
-    ...     output_size: int = 4
-
     >>> class MyModel(nn.Module, PyTorchModelHubMixin):
-    ...     def __init__(self, config: Config):
+    ...     def __init__(self, hidden_size: int = 512, vocab_size: int = 30000, output_size: int = 4):
     ...         super().__init__()
     ...         self.param = nn.Parameter(torch.rand(config.hidden_size, config.vocab_size))
     ...         self.linear = nn.Linear(config.output_size, config.vocab_size)
```
Comment (NielsRogge) [inline src/huggingface_hub/hub_mixin.py:492]:
```suggestion
    ...         self.param = nn.Parameter(torch.rand(hidden_size, vocab_size))
    ...         self.linear = nn.Linear(output_size, vocab_size)
```

File: docs/source/en/guides/integrations.md [hunk 2]
```diff
@@ -211,25 +209,15 @@ class PyTorchModelHubMixin(ModelHubMixin):
...
-         ),
-         model_summary=...,
-         model_type=...,
-         ...
-      )
-      (save_directory / "README.md").write_text(str(model))
-      torch.save(obj=self.module.state_dict(), f=save_directory / "pytorch_model.bin")
+    def _save_pretrained(self, save_directory: Path) -> None:
+        """Save weights from a Pytorch model to a local directory."""
+        save_model_as_safetensor(self.module, str(save_directory / SAFETENSORS_SINGLE_FILE))
```
Comment (LysandreJik) [inline docs/source/en/guides/integrations.md:219]: yess :tada:

File: src/huggingface_hub/hub_mixin.py [hunk 3]
```diff
@@ -246,32 +288,37 @@ def from_pretrained(
...
 
-            # Check if class expect a `config` argument
-            init_parameters = inspect.signature(cls.__init__).parameters
-            if "config" in init_parameters:
+            # Populate model_kwargs from config
+            for param in cls._init_parameters.values():
+                if param.name not in model_kwargs and param.name in config:
+                    model_kwargs[param.name] = config[param.name]
+
+            # Check if `config` argument is a dataclass
```
Comment (LysandreJik) [inline src/huggingface_hub/hub_mixin.py:302]:
```suggestion
            # Check if `config` argument was passed at init
```

File: src/huggingface_hub/utils/_typing.py [hunk 4]
```diff
@@ -14,10 +14,37 @@
...
+_JSON_SERIALIZABLE_TYPES = (int, float, str, bool, type(None))
+
+
+def is_jsonable(obj: Any) -> bool:
+    """Check if an object is JSON serializable.
+
+    This is a weak check, as it does not check for the actual JSON serialization, but only for the types of the object.
+    It works correctly for basic use cases but do not guarantee an exhaustive check.
+
+    Object is considered to be json serializable if:
```
Comment (LysandreJik) [inline src/huggingface_hub/utils/_typing.py:34]:
```suggestion
    Object is considered to be recursively json serializable if:
```

**Review by LysandreJik — APPROVED** [review]
Looks great! That is sure to come in handy with model serialization, no need to handle a config on the side anymore. Awesome!

**Comment by Wauplin** [comment]
Thanks for the review @LysandreJik :hugs: Yes I also expect it to make integrations much easier now :)

**Pull request merged**